code/modules/tgui_input/modal/modal.dm

```diff
@@ -0,0 +1,135 @@
+#define NULL_CLIENTMOB "Tgui modal loaded on a null client/mob"
+
+/**
+ * The tgui speech modal. This initializes an input window which hides until
+ * the user presses one of the speech hotkeys. Once an entry is set, it will
+ * delegate the speech to the proper channel.
+ */
+/datum/tgui_modal
+	/// The user who opened the window
+	var/client/client
+	/// Injury phrases to blurt out
+	var/list/hurt_phrases = list("GACK!", "GLORF!", "OOF!", "AUGH!", "OW!", "URGH!", "HRNK!")
+	/// Max message length
+	var/max_length = MAX_MESSAGE_LEN
+	/// The modal window
+	var/datum/tgui_window/window
+	/// Boolean for whether the tgui_modal was opened by the user.
+	var/window_open
+
+/** Assigned window to the client */
+/client/var/datum/tgui_modal/tgui_modal
+
+/** Creates the new input window to exist in the background. */
+/datum/tgui_modal/New(client/client, id)
+	src.client = client
+	window = new(client, id)
+	window.subscribe(src, .proc/on_message)
+	window.is_browser = TRUE
+
+/**
+ * Injects the scripts and styling into the window,
+ * then feeds it props for the chat channel and max message length.
+ */
+/datum/tgui_modal/proc/initialize()
+	set waitfor = FALSE
+	// Sleep to defer initialization to after client constructor
+	sleep(3 SECONDS)
+	window.initialize(
+			strict_mode = TRUE,
+			fancy = TRUE,
+			inline_css = file("tgui/public/tgui-modal.bundle.css"),
+			inline_js = file("tgui/public/tgui-modal.bundle.js"),
+	);
+
+/**
+ * Creates a JSON encoded message to open TGUI modals properly.
+ *
+ * Arguments:
+ * channel - The channel to open the modal in.
+ * Returns:
+ * string - A JSON encoded message to open the modal.
+ */
+/client/proc/tgui_modal_create_open_command(channel)
+	var/message = TGUI_CREATE_MESSAGE("open", list(
+		channel = channel,
+	))
+	return "\".output tgui_modal.browser:update [message]\""
+
+/**
+ * Ensures nothing funny is going on window load.
+ * Minimizes the winddow, sets max length, closes all
+ * typing and thinking indicators.
+ */
+/datum/tgui_modal/proc/load()
+	if(!client?.mob)
+		CRASH(NULL_CLIENTMOB)
+	window_open = FALSE
+	winset(client, "tgui_modal", "is-visible=false")
+	/// Sanity check in case the server ever changes MAX_LEN_MESSAGE
+	window.send_message("maxLength", list(
+		maxLength = max_length,
+	))
+	client?.mob?.cancel_thinking()
+	return TRUE
+
+/**
+ * Sets the window as "opened" server side, though it is already
+ * visible to the user. We do this to set local vars &&
+ * start typing (if enabled and in an IC channel). Logs the event.
+ *
+ * Arguments:
+ * payload - A list containing the channel the window was opened in.
+ */
+/datum/tgui_modal/proc/open(payload)
+	if(!client?.mob)
+		CRASH(NULL_CLIENTMOB)
```

Comment:
Here (and a few other places), don't we already expect a client to be using this (since it's TGUI)? Otherwise, can't we just check for !client and early return, instead of using ? everywhere

Reply:
very possible, i'm just being incredibly extra with regards to client